rids/network_capture.py

```diff
@@ -14,74 +14,112 @@
+  where server_hello may be `None` if only the client hello has been seen, and
+  will include both when the server_hello is found.  Each is a dict that has
+  `remote_ip`, `remote_port`, `sni` (if present), `ja3`, `ja3s` as properties.
+  """
+  def __init__(self, host_ip):
+    self._proc = _start_tshark([
+        '-f', 'tcp and not (src port 443 or dst port 443)',
+        '-Y', (f'(tls.handshake.type == 1 and ip.src == {host_ip})' +
+              f'or (tls.handshake.type == 2 and ip.dst == {host_ip})'),
+        '-Tfields',
+        # Ordering of the following fields determines the ordering in output,
+        # except that missing fields are skipped.
+        '-e', 'frame.time',
+        '-e', 'tcp.stream',
+        '-e', 'tls.handshake.type',
+        '-e', 'ip.src',
+        '-e', 'tcp.srcport',
+        '-e', 'ip.dst',
+        '-e', 'tcp.dstport',
+        '-e', 'tls.handshake.extensions_server_name',
+        '-e', 'tls.handshake.ja3',
+        '-e', 'tls.handshake.ja3_full',
+        '-e', 'tls.handshake.ja3s',
+        '-e', 'tls.handshake.ja3s_full',
+        '-l',
+    ])
+    self._tls_streams = {}
 
-    logging.basicConfig(level=logging.DEBUG,
-                        format='%(asctime)s %(levelname)-8s %(message)s',
-                        datefmt='%m-%d %H:%M',
-                        filename='/var/maldetector.log',
-                        filemode='a')
+  # Ignore these false positives.
+  # TODO: include these in handshake-related rules instead of filtering here?
+  _allowed_sni_port = set([
+      ('mtalk.google.com', 5228),
+      ('proxy-safebrowsing.googleapis.com', 80),
+      ('courier.push.apple.com', 5223),
+      ('imap.gmail.com', 993),
+  ])
 
-    for line in iter(input_stream, b''):
-        values = line.split('\t')
-        if len(values) >= 8:
-            stream_id = int(values[1])
-            if int(values[2]) == 1:  # client hello
-                sni_and_port = (values[7], int(values[6]))
-                if sni_and_port not in allowed_sni_port:
-                    unusual_tls_traffic[stream_id] = {
-                        'client_ts': values[0],
-                        'server_ip': values[5],
-                        'server_port': values[6],
-                        'server_name': values[7],
-                        'ja3': values[8],
-                        'ja3_full': values[9],
-                    }
-            elif int(values[2]) == 2:  # server hello
-                port = int(values[4])
-                if port == 443:
-                    continue
-                connection_details = unusual_tls_traffic.get(stream_id, None)
-                if not connection_details:
-                    continue
-                connection_details['ja3s'] = values[8]
-                connection_details['ja3s_full'] = values[9]
-                logging.info('connection_details %s', connection_details)
-                del unusual_tls_traffic[stream_id]
+  def scan(self):
```

Comment:
Make the `monitor` method return a `TlsConnectionInfo`.
Move the detection of suspicious TLS connections to a separate class that takes TlsConnectionInfo. The allowed endpoints should then be there.

That will allow us to test detection without tshark. And we can have different sources of connections, like a pcap file or even something from Outline.